domain/consensus/processes/consensusstatemanager/import_pruning_utxo_set.go

```diff
@@ -144,31 +127,60 @@ func (csm *consensusStateManager) commitSetPruningPointUTXOSetAll() error {
 		}
 	}
 
-	return dbTx.Commit()
-}
+	err = dbTx.Commit()
+	if err != nil {
+		return err
+	}
 
-type protoUTXOSetIterator struct {
-	utxoSet *utxoserialization.ProtoUTXOSet
-	index   int
+	return csm.importVirtualUTXOSetAndPruningPointUTXOSet()
 }
 
-func (p *protoUTXOSetIterator) Next() bool {
-	p.index++
-	return p.index < len(p.utxoSet.Utxos)
-}
+func (csm *consensusStateManager) importVirtualUTXOSetAndPruningPointUTXOSet() error {
+	onEnd := logger.LogAndMeasureExecutionTime(log, "importVirtualUTXOSetAndPruningPointUTXOSet")
+	defer onEnd()
 
-func (p *protoUTXOSetIterator) Get() (outpoint *externalapi.DomainOutpoint, utxoEntry externalapi.UTXOEntry, err error) {
-	entry, outpoint, err := utxo.DeserializeUTXO(p.utxoSet.Utxos[p.index].EntryOutpointPair)
+	log.Debugf("Starting to import virtual UTXO set and pruning point utxo set")
+	err := csm.consensusStateStore.StartImportingPruningPointUTXOSet()
 	if err != nil {
-		if serialization.IsMalformedError(err) {
-			return nil, nil, errors.Wrapf(ruleerrors.ErrMalformedUTXO, "malformed utxo: %s", err)
-		}
-		return nil, nil, err
+		return err
+	}
+
+	log.Debugf("Getting an iterator into the imported pruning point utxo set")
+	pruningPointUTXOSetIterator, err := csm.pruningStore.ImportedPruningPointUTXOIterator(csm.databaseContext)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Importing the virtual UTXO set")
+	err = csm.consensusStateStore.ImportPruningPointUTXOSetIntoVirtualUTXOSet(pruningPointUTXOSetIterator)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Importing the new pruning point UTXO set")
+	err = csm.pruningStore.CommitImportedPruningPointUTXOSet()
+	if err != nil {
+		return err
 	}
 
-	return outpoint, entry, nil
+	log.Debugf("Finishing to import virtual UTXO set and pruning point UTXO set")
+	return csm.consensusStateStore.FinishImportingPruningPointUTXOSet()
 }
 
-func protoUTXOSetToReadOnlyUTXOSetIterator(protoUTXOSet *utxoserialization.ProtoUTXOSet) model.ReadOnlyUTXOSetIterator {
-	return &protoUTXOSetIterator{utxoSet: protoUTXOSet, index: -1}
+func (csm *consensusStateManager) RecoverUTXOIfRequired() error {
+	hadStartedImportingPruningPointUTXOSet, err := csm.consensusStateStore.HadStartedImportingPruningPointUTXOSet()
+	if err != nil {
+		return err
+	}
+	if !hadStartedImportingPruningPointUTXOSet {
+		return nil
+	}
+
+	log.Warnf("Unimported pruning point UTXO set detected. Attempting to recover...")
+	err = csm.importVirtualUTXOSetAndPruningPointUTXOSet()
+	if err != nil {
+		return err
+	}
+	log.Warnf("Unimported UTXO set successfully recovered")
+	return nil
 }
```

Comment:
A. Shouldn't we also `ClearImportedPruningPointUTXOs` && `ClearImportedPruningPointMultiset` here?.
B. `importVirtualUTXOSetAndPruningPointUTXOSet` doesn't seem to validate the UTXO commitment which I think is needed here?

Reply:
A. Clearing is happening lower in the call stack, right after we call `RecoverUTXOIfRequired` itself
B. We may only have `HadStartedImportingPruningPointUTXOSet() == true` if we already had validated the commitment

Reply:
not sure I understood B